framework/calico_framework.py

```diff
@@ -0,0 +1,707 @@
+
+        Otherwise, this number will be seconds-since-epoch of the last
+        checkpoint.
+
+        For an unstarted test, this is set if/when
+        this test refuses to start while no other tests are running.
+        If time expires in this state, that means the mesos cluster was
+        unable to provide enough offers to launch this test when no other tests
+        were running.
+
+        For a started test, this value is reset at each task update, and
+        represents the last time
+        this Test heard from one of its tasks.
+        """
+
+        self.additional_info = None
+        """
+        Test-wide information which will be provided during the test output.
+        Useful for explaining a test-scope failure (vs. a task failure).
+        """
+
+        self.killed = False
+        """
+        Flag, when set to True, will indicate that this test is being shutdown.
+        """
+
+    def __repr__(self):
+        s_repr = "Test(%s" % self.name
+        if self.timeout is not None:
+            s_repr += (", timeout=%d" % (time.time() - self.timeout))
+        s_repr += ")"
+
+        return s_repr
+
+    @property
+    def complete(self):
+        """
+        True if the test is no longer running. This means killed tests too
+        """
+        return self.killed or self.tasks_with_state(
+            UNFINISHED_TASK_STATES) == []
+
+    def can_run_on(self, offers):
+        """
+        Returns true if the test can run on the offers passed in.
+        Offers passed in will be ignored if they are already reserved.
+        This function also reserves offers and plans tasks as it goes along.
+        It will rollback reservations if it realizes it can't finish
+        planning all tasks.
+        """
+        # First sort the tasks by slave
+        tasks_by_slave = {}
+        for task in [task for task in self.tasks if task.slave is not None]:
+            tasks_by_slave.setdefault(task.slave, []).append(task)
+
+        # Sort those groups by most tasks to least tasks
+        tasks_grouped_by_slave = sorted(tasks_by_slave.values(),
+                                        lambda x, y: len(y) - len(x))
+
+        # Quit early if there isn't enough unique slaves
+        if len(tasks_grouped_by_slave) > len(offers):
+            return False
+
+        # Then find the remaining tasks without a specified slave requirement
+        anywhere_tasks = [task for task in self.tasks if task.slave is None]
+
+        # Sort the offers by resources (this assumes each offer is from a
+        # unique slave
+        offers_by_tasks_launchable = sorted(offers, lambda x, y:
+        cmp(y.num_tasks_launchable, x.num_tasks_launchable))
+
+        # We now have the offers sorted by largest to smallest, and the
+        # tasks-per-slave grouped from largest to smallest.
+        try:
+            for offer, tasks in zip(offers_by_tasks_launchable,
+                                    tasks_by_slave.values()):
```

Comment:
Should be `tasks_grouped_by_slave`, otherwise you're not processing them in the sorted order.